transformer_lens/components.py

```diff
@@ -593,6 +598,33 @@ def forward(
             )
             / self.attn_scale
         )  # [batch, head_index, query_pos, key_pos]
+
+        # alibi encoding before applying causal mask
+        if self.cfg.positional_embedding_type == "alibi":
+            batch_size = attn_scores.size(0)
+            seq_len = attn_scores.size(-1)
+            additive_mask = torch.ones(batch_size, seq_len)
+            dtype = (
+                self.cfg.dtype
+                if self.cfg.dtype in [torch.float32, torch.float64]
+                else "torch.float32"
+            )
+            alibi = self.build_alibi_tensor(
+                attention_mask=additive_mask, num_heads=self.cfg.n_heads, dtype=dtype
+            ).to(attn_scores.device)
+
+            # Huggingface impl uses torch.Tensor.baddbmm, with alpha = 1/sqrt(d_head), and beta=1
+            # and alibi.baddbmm(q,k) = beta * alibi + alpha * (q@k),
+            # here the `attn_scores` is already scaled by a factor of self.attn_scale,
+            # we only need to add alibi matrix to the result
+            assert alibi.shape == (
+                attn_scores.size(0),
+                attn_scores.size(1),
+                1,
+                attn_scores.size(-1),
+            ), f"alibi shape {alibi.shape}, expecting {attn_scores.shape}"
```

Comment:
Also is this right? 

If the query shape is size 1 surelly this only works for predicting the next token, but for the
logits for all
previous tokens it would then give an incorrect answer? Would mean that training wouldn't work with
this code right (or any analysis that looks at logits other than from the last token )